### PR TITLE
[SPARK-15692][SQL]Improves the explain output of several physical plans by displaying embedded logical plan in tree style

### PR DESCRIPTION
**Pull request by clockfly**
## What changes were proposed in this pull request?

Improves the explain output of several physical plans by displaying embedded logical plan in tree style

Some physical plan contains a embedded logical plan, for example, `cache tableName query` maps to:

```
case class CacheTableCommand(
    tableName: String,
    plan: Option[LogicalPlan],
    isLazy: Boolean) 
  extends RunnableCommand
```

It is easier to read the explain output if we can display the `plan` in tree style.

**Before change:**

Everything is messed in one line.

```
scala> Seq((1,2)).toDF().createOrReplaceTempView("testView")
scala> spark.sql("cache table testView2 select * from testView").explain()
== Physical Plan ==
ExecutedCommand CacheTableCommand testView2, Some('Project [*]
+- 'UnresolvedRelation `testView`, None
), false
```

**After change:**

```
scala> spark.sql("cache table testView2 select * from testView").explain()
== Physical Plan ==
ExecutedCommand
:  +- CacheTableCommand testView2, false
:     :  +- 'Project [*]
:     :     +- 'UnresolvedRelation `testView`, None
```
## How was this patch tested?

Manual test.

### CODE REVIEW

**Comment by rxin** [comment]
This is pretty cool.

(I didn't actually review the pr but only read the description)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala [hunk 1]
```diff
@@ -264,7 +264,7 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
     expressions.flatMap(_.collect {case e: SubqueryExpression => e.plan.asInstanceOf[PlanType]})
   }
 
-  override def innerChildren: Seq[PlanType] = subqueries
+  override protected def innerChildren: Seq[TreeNode[_]] = subqueries
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: use `QueryPlan`? I don't think we will treat expression as child of query plan.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 2]
```diff
@@ -424,9 +424,12 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
    */
   protected def stringArgs: Iterator[Any] = productIterator
 
+  private def allChildren: Set[TreeNode[_]] = containsChild ++ innerChildren.toSet[TreeNode[_]]
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: use lazy val?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 3]
```diff
@@ -70,6 +71,8 @@ private[sql] case class InMemoryRelation(
     private[sql] var _batchStats: ListAccumulator[InternalRow] = null)
   extends logical.LeafNode with MultiInstanceRelation {
 
+  override protected def innerChildren: Seq[TreeNode[_]] = Seq(child) ++ super.innerChildren
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: Do we really need to call `super.innerChildren` here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala [hunk 4]
```diff
@@ -69,7 +72,7 @@ private[sql] case class ExecutedCommandExec(cmd: RunnableCommand) extends SparkP
     sqlContext.sparkContext.parallelize(sideEffectResult, 1)
   }
 
-  override def argString: String = cmd.toString
+  override def argString: String = ""
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/commands.scala]: why this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala (hunk 1 above)
Comment (clockfly) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala]: What is your question? This only changes the visibility and return type.

**Comment by clockfly** [comment]
@cloud-fan  updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala [hunk 5]
```diff
@@ -222,6 +225,8 @@ private[sql] case class InMemoryTableScanExec(
     @transient relation: InMemoryRelation)
   extends LeafExecNode {
 
+  override protected def innerChildren: Seq[TreeNode[_]] = Seq(relation) ++ super.innerChildren
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala]: ok seems this is the only special case, how about we add a comment to explain why we call `super.innerChildren` here? e.g. There are expressions inside this operator and may contain subquery expression.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 6]
```diff
@@ -424,9 +424,12 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
    */
   protected def stringArgs: Iterator[Any] = productIterator
 
+  private lazy val allChildren: Set[TreeNode[_]] = containsChild ++ innerChildren.toSet[TreeNode[_]]
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: `containsChild` looks pretty weird here (the name). I'd prefer `(children ++ innerChildren).toSet[TreeNode[_]]`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 7]
```diff
@@ -467,9 +470,11 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product {
   }
 
   /**
-   * All the nodes that are parts of this node, this is used by subquries.
+   * All the nodes that are private children of current node.
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala]: "Private children" is quite vague. Maybe:

> All the nodes that should be shown as a inner nested tree of this node. For example, this can be used to show sub-queries.

**Comment by liancheng** [comment]
LGTM except for two minor issues, and pending Jenkins.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala [hunk 8]
```diff
@@ -19,15 +19,19 @@ package org.apache.spark.sql.execution.command
...
 
-
 case class CacheTableCommand(
   tableName: String,
   plan: Option[LogicalPlan],
   isLazy: Boolean)
   extends RunnableCommand {
 
+  override protected def innerChildren: Seq[QueryPlan[_]] = {
+    plan.map(Seq(_)).getOrElse(Seq.empty)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/cache.scala]: how about `plan.toSeq`?

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by cloud-fan** [comment]
merging to master and 2.0!
